### ISSUE TITLE
Error in ClTracerLensing for certain step-function n(z)

### ISSUE THREAD

**Issue by tilmantroester** [LSSTDESC/CCL#350]
I'm getting an odd error when trying to create lensing tracers with certain n(z)'s. The affected n(z)'s aren't particularly special (they're step functions) but the occurrence of the error seems to be very sensitive to the specific n(z)'s.

The code snippet below is the shortest example I could reduce it too. The error usually occurs for upsample >~ 20. It works for upsample = 11, but doesn't work for 12, 13, works again for 14, but not 15. Using only parts of the n(z) makes it work again.

 
```
import numpy as np
import pyccl as ccl

def create_step_nofz(z, nofz, upsample=10):
    z_step = np.linspace(z[0], z[-1]+z[1]-z[0], len(nofz)*upsample, endpoint=False)
    nofz_step = np.zeros_like(z_step)
    for i in range(len(nofz)):
        nofz_step[i*upsample:(i+1)*upsample] = nofz[i]
    return z_step, nofz_step

z_coarse, nofz_coarse = np.loadtxt("./nofz_3.txt", unpack=True)

z_step, nofz_step = create_step_nofz(z_coarse, nofz_coarse, upsample=13)

ccl_cosmology = ccl.Cosmology(Omega_c=0.3,
                              Omega_b=0.05,
                              Omega_k=0.0,
                              h=0.7,
                              n_s=0.96,
                              sigma8=0.8)
lens = ccl.ClTracerLensing(ccl_cosmology, has_intrinsic_alignment=False,
                           n=[z_step, nofz_step],)

```

The somewhat unhelpful error message is

```
TypeErrorTraceback (most recent call last)
<ipython-input-1-59bee6104e9d> in <module>()
     23                               sigma8=0.8)
     24 lens = ccl.ClTracerLensing(ccl_cosmology, has_intrinsic_alignment=False,
---> 25                            n=[z_step, nofz_step],)

/Users/yooken/miniconda/envs/python3/lib/python3.5/site-packages/pyccl/cls.py in __init__(self, cosmo, has_intrinsic_alignment, n, z, bias_ia, f_red)
    193                  has_intrinsic_alignment=has_intrinsic_alignment,
    194                  z=z, n=n, bias=None, mag_bias=None,
--> 195                  bias_ia=bias_ia, f_red=f_red)
    196 
    197 

/Users/yooken/miniconda/envs/python3/lib/python3.5/site-packages/pyccl/cls.py in __init__(self, cosmo, tracer_type, has_rsd, has_magnification, has_intrinsic_alignment, z, n, bias, mag_bias, bias_ia, f_red, z_source)
     98                             self.z_rf, self.rf,
     99                             self.z_source,
--> 100                             status )
    101 
    102     def __del__(self):

TypeError: 'int' object is not iterable
```
[nofz_3.txt](https://github.com/LSSTDESC/CCL/files/1811917/nofz_3.txt)

Edit: I wrote one of the offensive n(z) to disk. The error persists when loading this n(z) again, so that it should be easier for anyone else to test.
[offensive_nofz.txt](https://github.com/LSSTDESC/CCL/files/1812038/offensive_nofz.txt)

**Comment by tilmantroester** [comment]
The issue is caused by a failure of the integration routine in window_lensing:
https://github.com/LSSTDESC/CCL/blob/f6f2d48e0a7653538a155cf02bf7982f53236eb6/src/ccl_cls.c#L64
This gives a roundoff error for certain values of chi for the offending n(z). I played around with different values for the number of Gauss-Kronrod points, the number of subdivisions, using the QAGS function, and the relative error. Only changing GSL_INTEG_GAUSS41 to GSL_INTEG_GAUSS61 and raising the error tolerance seems to get the integral to converge.

Would there be interest in exposing the GSL accuracy parameters similarly to how the spline parameters are exposed? That way users can choose higher accuracies if needed (which might resolve #347) or lower, if speed is of importance. 
If so, I could try to give this a shot.

**Comment by tilmantroester** [comment]
Addressed with PR #371.
